src/models/broker.cr

```diff
@@ -0,0 +1,64 @@
+require "rethinkdb-orm"
+
+require "./base/model"
+
+module PlaceOS::Model
+  class Broker < ModelBase
+    include RethinkORM::Timestamps
+    table :broker
+
+    enum AuthType
+      Certificate
+      NoAuth
+      UserPassword
+    end
+
+    enum_attribute auth_type : AuthType = ->{ AuthType::UserPassword }
+
+    attribute name : String
+    attribute description : String
+
+    attribute ip : String
+    attribute port : Int32
+    attribute tls : Bool = false
+
+    attribute username : String
+    attribute password : String
+
+    attribute certificate : String
+
+    # Regex filters for sensitive data.
+    # Matches will be replaced with a SHA256(match + organisation_id).
+    attribute filters : Array(String) = ->{ [] of String }
+
+    validate ->(this : Broker) {
+      return unless (filters = this.filters)
+      # Render regex errors
+      error_string = filters.each_with_object({} of String => String?) { |filter, e|
+        e[filter] = Regex.error?(filter)
+      }.compact.map { |f, e| "#{f} with '#{e}'" }.join(" and")
+
+      this.validation_error(:filters, error_string) unless error_string.empty?
+    }
+
+    def sanitize(scope : String, payload : String)
+      filters = self.filters
+      return payload if filters.nil? || filters.empty?
+
+      # TODO: Cache and reuse unless filters_changed?
+      regex = Regex.union(filters)
+
+      payload.gsub(regex) do |match, _|
+        sha256("#{scope}-#{match}")
+      end
```

Comment:
Naming here is potentially confusing. The first arg is the string, second arg is the match object.

```suggestion
      payload.gsub(regex) do |str|
        sha256("#{scope}-#{str}")
      end
```

More importantly, we probably _should_ use a HMAC here (as you suggested on chat).
If we want to keep this as a hash for performance reasons, switching this to `sha256("#{match}#{scope}")` will help protect against _some_ length extension attacks as scope is public (known length), but the match string will vary.

Use case wise - what's `scope`? We may whats being used as the key here to be constant throughout all messages. This will allow matching between entities referenced in different message (which is a key features for analytics), but without the info on who / what those filtered entities are.

Reply:
`scope`: the `organization_id` as defined in the spec. Steve made a point about supporting generic hierarchies, so I'm taking the top level key as `scope`

I'll rename `match` -> `string`, I believe your suggestion would be serializing a tuple of a `String` and a `MatchData` object.